### PR TITLE
refactor: use concrete error types for node resolution

### PR DESCRIPTION
**Pull request by dsherret**
This will help clean up some of the code in the CLI because we'll be able to tell how the resolution failed (not part of this PR).

### CODE REVIEW

File: ext/node/errors.rs [hunk 1]
```diff
@@ -1,170 +1,462 @@
...
+    impl<E> From<E> for $name
+    where
+      $kind_name: From<E>,
+    {
+      fn from(err: E) -> Self {
+        $name(Box::new($kind_name::from(err)))
+      }
+    }
+  };
 }
```
Comment (dsherret) [inline ext/node/errors.rs:39]: Not sure the best way to go about doing this, but I wanted all the error types to be boxed without much trouble (similar to this https://github.com/dtolnay/thiserror/issues/57#issuecomment-576437396), but still have the ability to inspect the errors. Using a macro here helped with this, but I feel like there must be something in the rust ecosystem that makes this easier.

File: ext/node/errors.rs [hunk 2]
```diff
@@ -1,170 +1,462 @@
...
+
+#[derive(Debug, Error)]
+pub enum UrlToNodeResolutionErrorKind {
+  #[error(transparent)]
+  TypeScriptNotSupported(#[from] TypeScriptNotSupportedInNpmError),
+  #[error(transparent)]
+  ClosestPkgJson(#[from] ClosestPkgJsonError),
+}
+
+// todo(THIS PR): this should be a TypeError
```
Comment (dsherret) [inline ext/node/errors.rs:217]: Need to figure out how to make these a TypeError.

File: ext/node/resolution.rs [hunk 3]
```diff
@@ -874,15 +874,15 @@ impl NodeResolver {
...
+            PackageTargetResolveErrorKind::InvalidPackageTarget(_) => {
+              last_error = Some(e);
               continue;
             }
-            return Err(last_error.unwrap());
-          }
+            _ => {
+              return Err(e);
+            }
+          },
```
Comment (dsherret) [inline ext/node/resolution.rs:885]: This is much better.

File: ext/node/errors.rs [hunk 4]
```diff
@@ -1,170 +1,462 @@
...
+  #[error(transparent)]
+  PackageSubpathResolve(#[from] PackageSubpathResolveError),
+  #[error(transparent)]
+  UrlToNodeResolution(#[from] UrlToNodeResolutionError),
+}
 
-  if let Some(message) = maybe_message {
-    msg = format!("{msg}. {message}");
-  }
+// todo(https://github.com/denoland/deno_core/issues/810): make this a TypeError
```
Comment (dsherret) [inline ext/node/errors.rs:54]: Note this todo, which is a purposeful regression introduced by this PR. I think it's an ok regression though that can be fixed once we make deno_core better.

Comment (bartlomieju) [reply ext/node/errors.rs:54]: Should be okay for now

**Review by bartlomieju — APPROVED** [review]
LGTM

File: cli/npm/managed/resolvers/local.rs [hunk 5]
```diff
@@ -696,6 +709,7 @@ fn junction_or_symlink_dir(
   old_path: &Path,
   new_path: &Path,
 ) -> Result<(), AnyError> {
+  use deno_core::anyhow::bail;
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/npm/managed/resolvers/local.rs:712]: It's not so useful atm.

**Pull request merged**